system/src/system_network_diagnostics.cpp

```diff
@@ -80,9 +80,28 @@ class SignalCache
 #endif
     }
 
+#if HAL_PLATFORM_CELLULAR
+    const CellularGlobalIdentity* getCGI()
+    {
+        system_tick_t m = millis();
+        if (ts_ == 0 || (m - ts_) >= NETWORK_INFO_CACHE_INTERVAL)
+        {
+            cgi_.size = sizeof(cgi_);
+            cgi_.version = CGI_VERSION_LATEST;
+            cellular_global_identity(&cgi_, nullptr);
+            ts_ = millis();
+        }
+        return &cgi_;
```

Comment:
I see this only updates every 1 second now, but couldn’t there still be a potential for a non-atomic update if MNC is just before 1 second and MCC is just after? I think based on how quickly these 4 parameters are updated it might even be easier to make the data non-atomic now since it would have forced the calling system to wait 1 full second between two reads of the same data set. 

Can we just force the original caller of the 4 get() functions to only be one call tot the new getCGI() function and remove the 1 second update mechanism?

Reply:
The result of an offline conversation _(scrum)_ was, as a team, we are pleased with the atomicity of the data in the `CellularGlobalIdentity` struct returned from the HAL layer. Building on that, we are confident that the four data points `mcc`, `mnc`, `lac` and `ci` will remain coherent, because all of the calls and caching in this PR are happening above the HAL layer and the time period begins at the first message invocation.

We did discuss the notion of a marker that could be used to indicate when a diagnostic group started emitting it's messages and when it was complete. This change could work for both signal strength and quality as well as cgi data. However, it was decided this change was beyond the scope of this PR, and would be best placed into the backlog.

Reply:
The key feature is that all the CGI diagnostics must be retrieved within 1 second of each other, which should be the case since they are registered within the same file.